q3 announce blog (#1567)
quarkusio/quarkusio.github.io: _posts/2022-11-07-road-to-quarkus-3.adoc

Especially for the core API’s we will aim to allow such migration to be trivial and thus can be automated in a similar manner as how we’ve done automatic conversion of Quarkus 2.x to use Jakarta named packages.

== Trying out Quarkus 3

For now the easiest way to get started is using https://code.quarkus.io/?S=io.quarkus.platform%3A3.0[code.quarkus.io] or use the Quarkus CLI:

```
quarkus app create --stream=3.0
```

Mind you that at this stage, while all the core extensions are available, only parts of the Quarkus platform is present. Not all extensions have yet migrated to Jakarta packages (e.g. Camel Quarkus or Kogito are not available).

-- TBD: mention the extensions already committed or planning to be ready.

== Upgrading to Quarkus 3

For existing applications where all extensions are available we have a early OpenRewrite recipe that you can try use.

There is a one-liner that attempts to do it automatically using a JBang script:

For Linux
```
curl -Ls https://sh.jbang.dev | bash -s - --fresh upgrade-to-quarkus3@quarkusio
```

For Windows
```

> iex "& { $(iwr https://ps.jbang.dev) } --fresh upgrade-to-quarkus@quarkusio"

Reviewer: Shouldn't this also be `upgrade-to-qurakus3@quarkusio`(it's missing the `3`)

Author: yup - fixed!

---
layout: post
title: 'Starting Quarkus 3'

> date: 2022-11-07

Reviewer 2: Once scheduled, this date needs to be changed (and the date in the filename too).

  - A *revamped dev UI*
  - <Your extension awesome feature/improvements here>

## Java Target 

Quarkus 3 continues (like Quarkus 2) to target minimal version of Java 11. The default recommendation for users will be to use Java 17. 

We will continue to revise minimal Java version based on user feedback and uptake of Java 17 in the ecosystem. For now we feel it is of higher value to allow as many Java 11 users as possible to upgrade to Quarkus 3 and use the Jakarta based dependencies. 

## Release Plan

We will continue the monthly cadence release of the 2.x version of Quarkus. Quarkus 3.x preview (alphas, betas) will be released in parallel. This blog marks the start of the first Quarkus 3 Alpha.

It will take more than a month to integrate the change set we want and collect feedback (from users, and extension maintainers...) from the previous release. Thus, 3.x release cadence will be a bit slower than usual.

Thus the intent is to have a time period of some months to allow continuous integration and start releasing a Quarkus Platform with the members that are ready to do and then continuously include more.

Proposed plan:

>   - Start doing 3.0.0.Alpha core and platform releases at regular intervals to allow easy integration builds, starting first from 2.13.3 in the coming week

Reviewer 2: It says `coming week`. Should we directly announce 3.0.0.Alpha1 or do you want a separate blog post with the Alpha1 release and the highlights of the Alpha1 release?

  - If all goes well, aim for a 3.0.0.Final in February 2023

== Breakage? Keep calm!

While the major version bump signals possible breaking changes, we still ask that all contributors still aim for a smooth upgrade experience. Changes like the move to Jakarta with its package renames have a big impact. We intend on doing what we can to make it so current extensions and user applications can migrate from Quarkus 2 to Quarkus 3 by providing tooling to perform the migration.

Especially for the core API’s we will aim to allow such migration to be trivial and thus can be automated in a similar manner as how we’ve done automatic conversion of Quarkus 2.x to use Jakarta named packages.

== Trying out Quarkus 3

For now the easiest way to get started is using https://code.quarkus.io/?S=io.quarkus.platform%3A3.0[code.quarkus.io] or use the Quarkus CLI:

```
quarkus app create --stream=3.0
```

Mind you that at this stage, while all the core extensions are available, only parts of the Quarkus platform is present. Not all extensions have yet migrated to Jakarta packages (e.g. Camel Quarkus or Kogito are not available).

> -- TBD: mention the extensions already committed or planning to be ready.

Reviewer 2: This probably needs to be removed?

Author: yeah - we have 3-4 to mention but better to mention in alpha2 where they had time to release against 3

== Upgrading to Quarkus 3

> For existing applications where all extensions are available we have a early OpenRewrite recipe that you can try use.

Reviewer 2: ```suggestion
For existing applications where all extensions are available we have a early OpenRewrite recipe that you can try.
```

There is a one-liner that attempts to do it automatically using a JBang script:

For Linux
```
curl -Ls https://sh.jbang.dev | bash -s - --fresh upgrade-to-quarkus3@quarkusio
```

For Windows
```
iex "& { $(iwr https://ps.jbang.dev) } --fresh upgrade-to-quarkus3@quarkusio"
```

You can also do it manually by downloading the https://raw.githubusercontent.com/quarkusio/quarkus/main/jakarta/quarkus3.yml[OpenRewrite recipe] and apply it manually with the following Maven command:

```
curl -o quarkus3.yml https://raw.githubusercontent.com/quarkusio/quarkus/main/jakarta/quarkus3.yml
mvn org.openrewrite.maven:rewrite-maven-plugin:4.36.0:run \
   -Drewrite.configLocation=quarkus3.yml \
   -DactiveRecipes=io.quarkus.openrewrite.Quarkus3
```

> Once this have been run your project should have its dependencies and source code updated to use Quarkus 3. 

Reviewer 2: ```suggestion
Once this has been run, your project should have its dependencies and source code updated to use Quarkus 3. 
```

> If it does not work for you it could be we missed something or you are using extensions not yet supporting Quarkus 3.

Reviewer 2: ```suggestion
If it does not work for you, it could be we missed something or you are using extensions not yet supporting Quarkus 3.
```

---
layout: post
title: 'Starting Quarkus 3'
date: 2022-11-07
tags: announcement
synopsis: 'Quarkus 3 is coming with new features and a move to Jakarta named APIs.'
author: maxandersen
---

Quarkus 3 will continue the tradition of propelling Java development forward by moving fast and bring a cutting-edge user experience with the latest technologies.

> Happy to have a first Quarkus 3 Alpha available for users to try out and extension writers to integrate with to get early feedback.

Reviewer 2: ```suggestion
We are happy to make a first Quarkus 3 Alpha available for users to try out and extension writers to integrate with so that we get early feedback.
```

> ## What is in it?

Reviewer 2: I think we should change this to the following as we don't have all of that in the first Alpha.
```suggestion
## What will be in it?
```

The current planned list of highlights for Quarkus 3 are:

>   - *Hibernate 6* - faster, safer, better (Check out https://www.youtube.com/watch?v=pc6QIwx0EL0[this presentation] to see how awesome it is).

Reviewer 2: ```suggestion
  - *Hibernate ORM 6* - faster, safer, better (Check out https://www.youtube.com/watch?v=pc6QIwx0EL0[this presentation] to see how awesome it is).
```

  - *Jakarta EE 10* API and packages - following the latest evolution of the Java world

>   - *Micro profile 6* - the latest update of all the MP specifications continuing evolving alongside the state of the practice

Reviewer 2: ```suggestion
  - *Eclipse MicroProfile 6* - the latest update of all the MP specifications continuing evolving alongside the state of the practice
```

  - *HTTP/3* support (aka *Quic*) - the https://www.redhat.com/architect/http3[new HTTP version] fixing the head-of-line blocking problem, using UDP.

>   - *IO_URING* - the next generation of asynchronous IO support. This is a complete game changer if you are interested in performance and efficiency (response time, latency…)

Reviewer 2: ```suggestion
  - *io_uring* - the next generation of asynchronous IO support. This is a complete game changer if you are interested in performance and efficiency (response time, latency…)
```

  - *Virtual threads* (aka Loom) and *Structured Concurrency Support* - based on the lessons learned from our https://www.redhat.com/architect/http3[initial integration], we are now looking into expanding and improving the support.

>   - *java.util.concurrent.Flow* - Time to evolve; Flow has been in the JDK since Java 9. Thus we want to move from Reactive Streams to java.util.concurrent.Flow.

Reviewer 2: ```suggestion
  - *java.util.concurrent.Flow* - Time to evolve; Flow has been in the JDK since Java 9. Thus we want to move from Reactive Streams to `java.util.concurrent.Flow`.
```

  - *Documentation* - Work started on better and more structured documentation following the https://diataxis.fr/[Diataxis] framework principles.
  - *Tooling for updates* - going to be a new major version, we want to make the migration as smooth as possible. Thus we will provide migration tools.
  - A new *gRPC server*, more flexible and easier to maintain
  - A *revamped dev UI*
  - <Your extension awesome feature/improvements here>

## Release Plan

We will continue the monthly cadence release of the 2.x version of Quarkus. Quarkus 3.x preview (alphas, betas) will be released in parallel. This blog marks the start of the first Quarkus 3 Alpha.

It will take more than a month to integrate the change set we want and collect feedback (from users, and extension maintainers...) from the previous release. Thus, 3.x release cadence will be a bit slower than usual.

Thus the intent is to have a time period of some months to allow continuous integration and start releasing a Quarkus Platform with the members that are ready to do and then continuously include more.

Proposed plan:

>   - Start doing 3.0.0.Alpha core and platform releases at regular intervals to allow easy integration builds, starting first from 2.13.3 in the coming week

Reviewer 2: So we will announce 3.0.0.Alpha1 separately? I think it kinda makes sense, just checking.
```suggestion
  - Start doing 3.0.0.Alpha core and platform releases at regular intervals to allow easy integration builds, starting first from 2.13.3 in the coming week
```

Author: Yeah, I was thinking of keeping 2.x and 3.x announce separate - possibly just link from one to the other if relevant.

  - 2.x releases will continue to happen on a monthly basis
  - Announce Quarkus 3 on Quarkus blog when 3.0.0.Alphas start being useful for users to try (This Blog!)

>   - Feedback from and Outreach to extension contributors and users on readiness and challenges to help and adjust

Reviewer 2: > Feedback from and Outreach

This is a bit weird?

  - If all goes well, aim for a 3.0.0.Final in February 2023

== Breakage? Keep calm!

While the major version bump signals possible breaking changes, we still ask that all contributors still aim for a smooth upgrade experience. Changes like the move to Jakarta with its package renames have a big impact. We intend on doing what we can to make it so current extensions and user applications can migrate from Quarkus 2 to Quarkus 3 by providing tooling to perform the migration.

Especially for the core API’s we will aim to allow such migration to be trivial and thus can be automated in a similar manner as how we’ve done automatic conversion of Quarkus 2.x to use Jakarta named packages.

== Trying it out

> For now the easiest way to get started is using https://code.quarkus.io/?S=io.quarkus.platform%3A3.0[code.quarkus.io] or use the Quarkus cli:

Reviewer 2: ```suggestion
For now the easiest way to get started is using https://code.quarkus.io/?S=io.quarkus.platform%3A3.0[code.quarkus.io] or use the Quarkus CLI:
```

```
quarkus app create --stream=3.0
```

> Mind you that at this stage only parts of the Quarkus platform is present. Not all extensions have yet migrated to Jakarta packages.

Reviewer 2: ```suggestion
Mind you that at this stage, while all the core extensions are available, only parts of the Quarkus platform is present. Not all extensions have yet migrated to Jakarta packages (e.g. Camel Quarkus or Kogito are not available).
```

-- TBD: mention the extensions already committed or planning to be ready.

> For existing applications where all extensions are available try migrate from using `javax.` to `jakarta.` manually and report back if it works or not for you.

Reviewer 2: Maybe add:
```
You can automate the transformation with the Eclipse Jakarta Transformer.
```
I think people should probably use this until we push our own tool because not all `javax` packages have changed. So if you do a search/replace for everything javax it will fail.

We will provide tooling to do some if not most of the necessary migration - we will mention that in migration guides and release blogs when available.

> -- TBD: is there a openrewrite (or eclipse transformer) command/call we can offer to get people partial there?

Reviewer 2: So yes, we need that, I'll have a look next week.
